[Merged by Bors] - refactor: make library notes a definition

grunweg opened the pull request:
Make library notes a definition of `Unit` type, whose doc-string is the note's content. As a consequence,
- doc-gen will display all library notes without any custom logic,
- library notes are shown upon hover (via the use of the doc-string),
- go to definition for library notes is easy

Automate this using a macro, which also enforces that all library notes are inside the Mathlib.LibraryNote namespace.
*Temporarily*, we use the `library_note2` name; if/when the batteries version is changed, we can reclaim the original name.

This PR does not address how to *refer* to library notes nicely (in a way that is shown in the documentation, checked for typos, with sufficient imports ensured etc.). That is left to a future PR.

This is proposal 2a from this zulip discussion: [#mathlib4 > library notes @ 💬](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/library.20notes/near/510614740)

------------

Open questions/TODOs:
- should this go in batteries instead? can the old definition just be overwritten?
- otherwise, what's the best place for the `LibraryNote` definition? A single file just for this feels overkill; I put it in Mathlib.Tactic.Basic for now. I could use Mathlib.Init, of course (but that might be excessive)


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg commented on Mathlib/Algebra/Group/Pointwise/Set/Scalar.lean:
```diff
@@ -49,18 +49,6 @@ pointwise subtraction
-Because the pointwise action can easily be spelled out in such cases, we give higher priority to the
-nat and int actions.
--/
-
```
This note was duplicated: I have chosen to delete this copy.

adomani commented on Mathlib/Tactic/Basic.lean:
```diff
@@ -160,3 +160,16 @@ elab (name := clearValue) "clear_value" hs:(ppSpace colGt term:max)+ : tactic =>
+
+-- XXX: how to add the doc-string into the generated macro?
+macro "library_note" dc:docComment name:name : command =>
+  `(def $(mkIdent (Name.append `LibraryNote name.getName)) : LibraryNote := ())
```
Or maybe even
```lean
macro dc:docComment " library_note " name:ident : command =>
  `($dc:docComment def $(mkIdent (Name.append `LibraryNote name.getId)) : LibraryNote := ())
```

grunweg committed:
feat: make library notes a definition,
and add a macro to conveniently create it

grunweg committed:
chore: convert all but one library note

grunweg committed:
Remove duplicate library note

grunweg committed:
Namespace

grunweg committed:
Fix

adomani commented on Mathlib/Tactic/Basic.lean:
```diff
@@ -160,3 +160,9 @@ elab (name := clearValue) "clear_value" hs:(ppSpace colGt term:max)+ : tactic =>
+/-- A mathlib library note: the note's content should be contained in its doc-string. -/
+def LibraryNote := Unit
+
+open Lean in macro "library_note2 " ppSpace name:ident dc:docComment : command =>
```
```suggestion
open Lean in macro "library_note2 " name:ident ppSpace dc:docComment : command =>
```
I think that this will pretty print better.

grunweg committed:
Update Mathlib/Tactic/Basic.lean

Co-authored-by: damiano <adomani@gmail.com>

grunweg committed:
Fix syntax error; add doc-string

grunweg committed:
Merge branch 'master' into MR-librarynote-def

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

Vierkantor reviewed (commented):
If we merge this PR (which I am in favour of), we would also need to rename all the references to the notes that got renamed. I'll merge master and start doing so.

Vierkantor committed:
Merge remote-tracking branch 'upstream/master' into MR-librarynote-def

Vierkantor committed:
Update note [SetLike Aesop ruleset] to library_note2

Vierkantor committed:
Fix references to `partiallyAppliedExtLemmas` by actually porting that note.

The note [partially-applied ext lemmas] never ended up being ported to Mathlib 4 since it lived in `tactic/ext.lean`, which is now a part of core Lean. So port the note and update the references to it.

Vierkantor committed:
Search and replace old library notes with the new ones.

This was accomplished by doing a long sequence of searches-and-replaces in all **/*.lean files. To verify this covers all our notes, I searched for "see note [...]" and made sure there were no spaces in between the brackets; I also scrolled through the search results for a visual inspection.

This PR also makes the references consistent: "{S,s}ee {library,} {N,n}ote [...]" becomes "{S,s}ee note [...]".

Vierkantor committed:
Merge remote-tracking branch 'upstream/master' into MR-librarynote-def

grunweg commented:
French quotes, I love it! (Sorry to hear that you only realised this now...)

grunweg commented:
I spot-checked perhaps 10% of the mechanical changes (and still agree with the conceptual changes). If you'd like to review the conceptual changes, perhaps between you and me that's enough to maintainer merge? What do you think?

Vierkantor committed:
Revert "Search and replace old library notes with the new ones."

This reverts commit 90dd7d810ccc936acc8abdaf9a0437e2f8422ad4.

Vierkantor committed:
Use french quotes to quote the identifier

grunweg commented:
Let me approve your part of the change: the original change (making library_note a definition) should be reviewed by somebody else, of course.
maintainer delegate

Vierkantor committed:
Fix name of out-param inheritance too

Co-authored-by: Michael Rothgang <rothgami@math.hu-berlin.de>

jcommelin reviewed (commented):
bors d+

Vierkantor committed:
Mention french quotes in the docstring.

I had to commit this manually because GitHub doesn't allow me to type or copy/paste the french quotes?

Vierkantor committed:
Delete double whitespace

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>
